internal/topology/variables/cluster_variable_defaulting.go

```diff
@@ -86,83 +81,143 @@ func defaultClusterVariables(clusterVariables []clusterv1.ClusterVariable, clust
 		Type: "object",
 		Properties: map[string]apiextensions.JSONSchemaProps{
-			clusterClassVariable.Name: *apiExtensionsSchema,
+			definition.Name: *apiExtensionsSchema,
 		},
 	}
 
 	// Default the variable via the structural schema library.
 	ss, err := structuralschema.NewStructural(wrappedSchema)
 	if err != nil {
 		return nil, field.ErrorList{field.Invalid(fldPath, "",
-			fmt.Sprintf("failed defaulting variable %q: %v", clusterVariable.Name, err))}
+			fmt.Sprintf("failed defaulting variable %q: %v", currentValue.Name, err))}
 	}
 	structuraldefaulting.Default(wrappedVariable, ss)
 
 	// Marshal the defaulted value.
-	defaultedVariableValue, err := json.Marshal(wrappedVariable[clusterClassVariable.Name])
+	defaultedVariableValue, err := json.Marshal(wrappedVariable[definition.Name])
 	if err != nil {
 		return nil, field.ErrorList{field.Invalid(fldPath, "",
-			fmt.Sprintf("failed to marshal default value of variable %q: %v", clusterClassVariable.Name, err))}
+			fmt.Sprintf("failed to marshal default value of variable %q: %v", definition.Name, err))}
 	}
-	return &clusterv1.ClusterVariable{
-		Name: clusterClassVariable.Name,
+	v := &clusterv1.ClusterVariable{
+		Name: definition.Name,
 		Value: apiextensionsv1.JSON{
 			Raw: defaultedVariableValue,
 		},
-	}, nil
+		DefinitionFrom: definition.From,
+	}
+
+	return v, nil
+}
+
+// getAllVariables returns a correctly ordered list of all variables defined in the ClusterClass and the Cluster.
+func getAllVariables(values []clusterv1.ClusterVariable, valuesIndex map[string]map[string]clusterv1.ClusterVariable, definitions []clusterv1.ClusterClassStatusVariable) []clusterv1.ClusterVariable {
+	// allVariables is used to get a full correctly ordered list of variables.
+	allVariables := []clusterv1.ClusterVariable{}
+	uniqueVariableDefinitions := map[string]bool{}
+
+	// Add any values that already exist.
+	allVariables = append(allVariables, values...)
+
+	// Add variables from the ClusterClass, which currently don't exist on the Cluster.
+	for _, variable := range definitions {
+		for _, definition := range variable.Definitions {
+			definitionFrom := definition.From
+
+			// 1) If there is a value in the Cluster with this definitionFrom or with an empty definitionFrom this variable does not need to be defaulted.
+			if _, ok := valuesIndex[variable.Name]; ok {
+				if _, ok := valuesIndex[variable.Name][definitionFrom]; ok {
+					continue
+				}
+				if _, ok := valuesIndex[variable.Name][emptyDefinitionFrom]; ok {
+					continue
+				}
+			}
+
+			// 2) If the definition has no conflicts no variable of the same name is defined in the Cluster set the definitionFrom to emptyDefinitionFrom.
+			if !variable.DefinitionsConflict && len(valuesIndex[variable.Name]) == 0 {
+				definitionFrom = emptyDefinitionFrom
+			}
+
+			// 3) If a variable with this name and definition has been added already, continue.
+			if _, ok := uniqueVariableDefinitions[definitionFrom+variable.Name]; ok {
```

Comment:
Let's add a godoc comment in which case we are hitting this if
(I think it's the case where we have multiple definitions with no conflict and no value in the Cluster spec, which is roughly when we hit case 2. in a scenario where we have multiple definitions (?))